Classes/Private/Manager/Objects/OCTActiveFile.m

```diff
@@ -0,0 +1,530 @@
+//
+//  OCTActiveFile.m
+//  objcTox
+//
+//  Created by stal on 9/7/2015.
+//  Copyright © 2015 Zodiac Labs. All rights reserved.
+//
+
+#import "DDLog.h"
+#import "OCTTox.h"
+#import "OCTActiveFile+Variants.h"
+#import "OCTSubmanagerFiles.h"
+#import "OCTMessageFile.h"
+#import "OCTSubmanagerFiles+Private.h"
+#import "OCTSubmanagerObjects+Private.h"
+#import "OCTRealmManager.h"
+
+#include <mach/clock.h>
+#include <mach/mach_host.h>
+#include <mach/mach_port.h>
+
+#undef LOG_LEVEL_DEF
+#define LOG_LEVEL_DEF LOG_LEVEL_DEBUG
+
+/* these are annoying, and because we get all file numbers from toxcore,
+ * casting will never truncate them because they were never longs to begin
+ * with. */
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wshorten-64-to-32"
+
+static time_t OCTGetSystemUptime(void)
+{
+    clock_serv_t muhclock;
+    mach_timespec_t machtime;
+
+    host_get_clock_service(mach_host_self(), SYSTEM_CLOCK, &muhclock);
+    clock_get_time(muhclock, &machtime);
+    mach_port_deallocate(mach_task_self(), muhclock);
+
+    return machtime.tv_sec;
+}
+
+@implementation OCTActiveFile
+
+@dynamic estimatedTimeRemaining;
+@dynamic progress;
+@dynamic bytesPerSecond;
+
+- (instancetype)init
+{
+    self = [super init];
+    if (! self) {
+        return nil;
+    }
+
+    self.transferRateCounters = calloc(sizeof(OCTToxFileSize), AVERAGE_SECONDS);
+    for (int i = 0; i < AVERAGE_SECONDS; ++i) {
+        self.transferRateCounters[i] = -1;
+    }
+
+    return self;
+}
+
+- (void)dealloc
+{
+    DDLogDebug(@"OCTActiveFile dealloc");
+    free(self.transferRateCounters);
+}
+
+#pragma mark - Speed counting stuff
+
+- (void)incrementRollingIndex:(long)n
+{
+    @synchronized(self) {
+        for (int i = 0; i < n; ++i) {
+            self.rollingIndex = (self.rollingIndex + 1) % AVERAGE_SECONDS;
+            self.transferRateCounters[self.rollingIndex] = 0;
+        }
+    }
+}
+
+- (void)countBytes:(NSUInteger)size
+{
+    self.bytesMoved += size;
+
+    time_t now = OCTGetSystemUptime();
+    time_t delta = now - self.lastCountedTime;
+    self.lastCountedTime = now;
+
+    NSAssert(delta >= 0, @"Detected a temporal anomaly. objcTox currently does not support Tox FTL, nor phone-microwave-"
+             "based time travel. Please file a bug.");
+
+    if (delta != 0) {
+        [self incrementRollingIndex:delta];
+    }
+
+    self.transferRateCounters[self.rollingIndex] += size;
```

Comment:
It's probably okay. I think a mutex is too heavy-handed for a method that gets called so often. (a rough calculation `6 MB/s -> 6,000,000 bytes in ~1430 byte packets -> 4,195.8 message sends/s, yikes!`, and that's just this method)

Since this is only ever set from the Tox queue, the worst that could happen is the client gets a wrong estimated time value (and we exclude the current hot-second from the average) which will be wiped in the next progress update.